mlflow/langchain/__init__.py

```diff
@@ -328,6 +429,48 @@ def log_model(
             type(lc_model.agent.llm_chain.llm).__name__,
         )
 
+    if isinstance(
+        lc_model,
+        (
+            RetrievalQA,
+            APIChain,
+            HypotheticalDocumentEmbedder,
+            SQLDatabaseChain,
+        ),
+    ):
+        if isinstance(lc_model, RetrievalQA) and version.parse(
+            langchain.__version__
+        ) < version.parse("0.0.194"):
+            raise mlflow.MlflowException.invalid_parameter_value(
+                _UNSUPPORTED_LANGCHAIN_VERSION_ERROR_MESSAGE.format(
+                    instnace_type=type(lc_model).__name__
+                )
+            )
+        if loader_fn is None:
+            raise mlflow.MlflowException.invalid_parameter_value(
+                "For {instnace_type} models, a `loader_fn` must be provided.".format(
+                    instnace_type=type(lc_model).__name__
+                )
+            )
+        if not isinstance(loader_fn, types.FunctionType):
+            raise mlflow.MlflowException.invalid_parameter_value(
+                "The `loader_fn` must be a function that retruns a {unserializable_object}.".format(
+                    unserializable_object=unserializable_object_name_map[type(lc_model).__name__]
+                )
+            )
+
+    # infer signature if signature is not provided
+    if signature is None:
+        input_columns = [
+            ColSpec(type=DataType.string, name=input_key) for input_key in lc_model.input_keys
+        ]
+        input_schema = Schema(input_columns)
+        output_columns = [
+            ColSpec(type=DataType.string, name=output_key) for output_key in lc_model.output_keys
+        ]
+        output_schema = Schema(output_columns)
+        signature = ModelSignature(input_schema, output_schema)
```

Comment:
I moved the model signature from `save_model()` to `log_model()`. If we infer signature in `save_model`, then `log_model` passes a `None` signature to `Model.log()` which calls `save_model()`, and `Model.log()` will print a warning complaining about missing signature, even though the signature is later inferred in `save_model()`.
https://github.com/mlflow/mlflow/blob/fd45110ba88ae68c1af425b9e26046a91f14c845/mlflow/models/model.py#L575-L578